// nativescript-ng2-carousel/nativescript-ng2-carousel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var enums_1 = require("ui/enums");
var image_1 = require("ui/image");
var stack_layout_1 = require("ui/layouts/stack-layout");
var grid_layout_1 = require("ui/layouts/grid-layout");
var label_1 = require("ui/label");
var gestures_1 = require("ui/gestures");
var image_source_1 = require("image-source");
var image_source_2 = require("image-source");
var CarouselDirective = /** @class */ (function () {
    function CarouselDirective(elem) {
        this.elem = elem;
        this.totalItems = 0;
        this.arrowType = CarouselArrowTypes.NORMAL;
        // Private control attributes
        this.direction = null;
        this.currentImage = 0;
        this.movingImages = false;
        this.indexMoveLeft = null;
        this.indexMoveRight = null;
        this.indexMoveCenter = null;
        this.container = elem.nativeElement;
    }
    CarouselDirective_1 = CarouselDirective;
    CarouselDirective.prototype.ngAfterViewInit = function () {
        this.initOptions();
        this.initContainer();
        this.initImagesLayout();
        this.initSlides();
        this.initControls();
        this.initAutoPlay();
    };
    /**
     * Get and set options from directive
     */
    CarouselDirective.prototype.initOptions = function () {
        // Animation duration (in ms)
        if (this.carouselAnimationSpeed && CarouselDirective_1.isNumeric(this.carouselAnimationSpeed)) {
            this.carouselAnimationSpeed = +this.carouselAnimationSpeed;
        }
        else {
            this.carouselAnimationSpeed = CarouselDirective_1.animationSpeedDefault;
        }
        // Autoplay (in ms) + animation duration
        if (this.carouselSpeed && CarouselDirective_1.isNumeric(this.carouselSpeed)) {
            this.carouselSpeed = +(this.carouselSpeed);
        }
        else {
            this.carouselSpeed = CarouselDirective_1.autoPlaySpeedDefault;
        }
        // Set arrow type
        if (this.carouselArrows) {
            switch (this.carouselArrows) {
                case 'none':
                    this.arrowType = CarouselArrowTypes.NONE;
                    break;
                case 'small':
                    this.arrowType = CarouselArrowTypes.SMALL;
                    break;
                case 'normal':
                    this.arrowType = CarouselArrowTypes.NORMAL;
                    break;
                case 'bold':
                    this.arrowType = CarouselArrowTypes.BOLD;
                    break;
                case 'narrow':
                    this.arrowType = CarouselArrowTypes.NARROW;
                    break;
            }
        }
    };
    /**
     * Init carousel layout
     */
    CarouselDirective.prototype.initContainer = function () {
        this.container.horizontalAlignment = "center";
        this.container.addRow(new grid_layout_1.ItemSpec(1, "auto"));
        this.container.addColumn(new grid_layout_1.ItemSpec(1, "star"));
        this.container.addColumn(new grid_layout_1.ItemSpec(1, "star"));
    };
    /**
     * Init sliders layout
     */
    CarouselDirective.prototype.initImagesLayout = function () {
        this.totalItems = this.carousel.length;
        this.carouselSlides = new grid_layout_1.GridLayout();
        grid_layout_1.GridLayout.setColumnSpan(this.carouselSlides, 2);
        this.container.addChild(this.carouselSlides);
    };
    /**
     * Init carousel sliders provided in "carousel" directive attribute
     */
    CarouselDirective.prototype.initSlides = function () {
        var _this = this;
        this.carousel.forEach(function (slide, i) {
            var gridLayout = new grid_layout_1.GridLayout();
            gridLayout.addRow(new grid_layout_1.ItemSpec(1, "auto"));
            gridLayout.visibility = i == 0 ? "visible" : "collapse";
            if (slide.url) {
                var image = CarouselDirective_1.generateImageSliderFromUrl(slide.url);
                gridLayout.addChild(image);
            }
            if (slide.file && slide.file.indexOf('res://') !== 0) {
                var image = CarouselDirective_1.generateImageSliderFromFile(slide.file);
                gridLayout.addChild(image);
            }
            if (slide.file && slide.file.indexOf('res://') === 0) {
                var image = CarouselDirective_1.generateImageSliderFromResource(slide.file);
                gridLayout.addChild(image);
            }
            if (slide.title) {
                var title = CarouselDirective_1.generateTitleSlider(slide.title);
                if (_this.carouselLabelOverlay) {
                    gridLayout.addRow(new grid_layout_1.ItemSpec(1, "auto"));
                    grid_layout_1.GridLayout.setRow(title, 1);
                }
                gridLayout.addChild(title);
            }
            _this.carouselSlides.addChild(gridLayout);
        });
    };
    /**
     * Load images from URL
     * @param src
     * @returns {Image}
     */
    CarouselDirective.generateImageSliderFromUrl = function (src) {
        var image = new image_1.Image();
        image.src = src;
        image.className = "slider-image";
        return image;
    };
    /**
     * Load images from file
     * @param path
     * @returns {Image}
     */
    CarouselDirective.generateImageSliderFromFile = function (path) {
        var image = new image_1.Image();
        image.imageSource = image_source_1.fromFile(path);
        image.className = "slider-image";
        return image;
    };
    /**
     * Load images from file
     * @param path
     * @returns {Image}
     */
    CarouselDirective.generateImageSliderFromResource = function (path) {
        var image = new image_1.Image();
        var pathRaw = path.replace('res://', '');
        image.imageSource = image_source_2.fromResource(pathRaw);
        image.className = "slider-image";
        return image;
    };
    /**
     * Generate title slider element
     * @param title
     * @returns {Label}
     */
    CarouselDirective.generateTitleSlider = function (title) {
        var label = new label_1.Label();
        label.text = title;
        label.textWrap = true;
        label.className = 'slider-title';
        return label;
    };
    /**
     * Init carousel controls
     */
    CarouselDirective.prototype.initControls = function () {
        var _this = this;
        if (this.totalItems > 1) {
            // Get Arrow type
            var arrowType = this.getArrowType();
            // Left arrow label
            var lLabel = new label_1.Label();
            lLabel.text = String.fromCharCode(parseInt(arrowType.l, 16));
            // Left arrow layout
            var lStackLayout = new stack_layout_1.StackLayout();
            lStackLayout.addChild(lLabel);
            lStackLayout.horizontalAlignment = "left";
            lStackLayout.on(gestures_1.GestureTypes.tap, function () {
                _this.stopStartAutoplay();
                _this.swipe(CarouselDirections.DIRECTION_LEFT);
            });
            lStackLayout.className = 'arrow arrow-left';
            grid_layout_1.GridLayout.setColumn(lStackLayout, 0);
            this.container.addChild(lStackLayout);
            // Right arrow label
            var rLabel = new label_1.Label();
            rLabel.text = String.fromCharCode(parseInt(arrowType.r, 16));
            // Left arrow layout
            var rStackLayout = new stack_layout_1.StackLayout();
            rStackLayout.addChild(rLabel);
            rStackLayout.horizontalAlignment = "right";
            rStackLayout.on(gestures_1.GestureTypes.tap, function () {
                _this.stopStartAutoplay();
                _this.swipe(CarouselDirections.DIRECTION_RIGHT);
            });
            rStackLayout.className = 'arrow arrow-right';
            grid_layout_1.GridLayout.setColumn(rStackLayout, 1);
            this.container.addChild(rStackLayout);
        }
    };
    /**
     * Init caroussel autoplay
     */
    CarouselDirective.prototype.initAutoPlay = function () {
        var _this = this;
        if (this.carouselSpeed && CarouselDirective_1.isNumeric(this.carouselSpeed)) {
            clearInterval(this.autoPlayIntervalId);
            this.autoPlayIntervalId = setInterval(function () {
                _this.swipe(CarouselDirections.DIRECTION_RIGHT);
            }, this.carouselSpeed + this.carouselAnimationSpeed);
        }
    };
    /**
     * Stop on gesture detected, resume after 4 seconds
     */
    CarouselDirective.prototype.stopStartAutoplay = function () {
        var _this = this;
        if (this.autoPlayIntervalId) {
            clearTimeout(this.autoPlayTimeoutId);
            clearInterval(this.autoPlayIntervalId);
            this.autoPlayTimeoutId = setTimeout(function () {
                _this.swipe(CarouselDirections.DIRECTION_RIGHT);
                _this.initAutoPlay();
            }, 4000);
        }
    };
    /**
     * Animate right to left or left to right
     * @param direction
     * @returns {boolean}
     */
    CarouselDirective.prototype.swipe = function (direction) {
        var _this = this;
        // Do nothing, hay solo una imagen...
        if (this.totalItems < 2 || this.movingImages) {
            return false;
        }
        // Animate slides
        this.direction = direction;
        this.movingImages = true;
        this.setDirectionValues();
        this.animateSlides();
        // Reset all after animation
        setTimeout(function () { return _this.resetAnimationValues(); }, this.carouselAnimationSpeed);
    };
    /**
     * Animate slides
     */
    CarouselDirective.prototype.animateSlides = function () {
        for (var i = 0; i < this.carouselSlides.getChildrenCount(); i++) {
            // Get view
            var view = this.carouselSlides.getChildAt(i);
            // Get element width + image visibility
            var elementWidth = this.elem.nativeElement.getActualSize().width;
            view.visibility = [this.indexMoveCenter, this.indexMoveLeft, this.indexMoveRight].indexOf(i) > -1 ? "visible" : "collapse";
            // Perfrom animation
            this.checkCL(view, i, elementWidth);
            this.checkCR(view, i, elementWidth);
            this.checkRC(view, i, elementWidth);
            this.checkLC(view, i, elementWidth);
        }
    };
    /**
     * Move image center -> left
     * @param view
     * @param index
     * @param elementWidth
     */
    CarouselDirective.prototype.checkCL = function (view, index, elementWidth) {
        if (this.indexMoveLeft == index) {
            view.translateX = 0;
            view.animate({
                translate: { x: elementWidth, y: 0 },
                duration: this.carouselAnimationSpeed,
                curve: enums_1.AnimationCurve.easeIn
            });
        }
    };
    /**
     * Move image right -> center
     * @param view
     * @param index
     * @param elementWidth
     */
    CarouselDirective.prototype.checkRC = function (view, index, elementWidth) {
        if (this.indexMoveCenter == index && this.direction == CarouselDirections.DIRECTION_LEFT) {
            view.translateX = -elementWidth;
            view.animate({
                translate: { x: 0, y: 0 },
                duration: this.carouselAnimationSpeed,
                curve: enums_1.AnimationCurve.easeOut
            });
        }
    };
    /**
     * Move image center -> right
     * @param view
     * @param index
     * @param elementWidth
     */
    CarouselDirective.prototype.checkCR = function (view, index, elementWidth) {
        if (this.indexMoveRight == index) {
            view.translateX = 0;
            view.animate({
                translate: { x: -elementWidth, y: 0 },
                duration: this.carouselAnimationSpeed,
                curve: enums_1.AnimationCurve.easeIn
            });
        }
    };
    /**
     * Move image left -> center
     * @param view
     * @param index
     * @param elementWidth
     */
    CarouselDirective.prototype.checkLC = function (view, index, elementWidth) {
        if (this.indexMoveCenter == index && this.direction == CarouselDirections.DIRECTION_RIGHT) {
            view.translateX = elementWidth;
            view.animate({
                translate: { x: 0, y: 0 },
                duration: this.carouselAnimationSpeed,
                curve: enums_1.AnimationCurve.easeOut
            });
        }
    };
    /**
     * Set values to perform the animation
     */
    CarouselDirective.prototype.setDirectionValues = function () {
        switch (this.direction) {
            // right to left
            case CarouselDirections.DIRECTION_LEFT:
                this.indexMoveLeft = this.currentImage;
                this.currentImage = ((this.currentImage == 0 ? this.totalItems : this.currentImage) - 1) % this.totalItems;
                this.indexMoveCenter = this.currentImage;
                break;
            // left to right
            case CarouselDirections.DIRECTION_RIGHT:
                this.indexMoveRight = this.currentImage;
                this.currentImage = (this.currentImage + 1) % this.totalItems;
                this.indexMoveCenter = this.currentImage;
                break;
        }
    };
    /**
     * Reset values after animation
     */
    CarouselDirective.prototype.resetAnimationValues = function () {
        this.indexMoveLeft = null;
        this.indexMoveRight = null;
        this.indexMoveCenter = null;
        this.movingImages = false;
    };
    /**
     * Get arrow type to be displayed in frontend
     * @returns {{l: string, r: string}}
     */
    CarouselDirective.prototype.getArrowType = function () {
        var ret = { l: '', r: '' };
        switch (this.arrowType) {
            case CarouselArrowTypes.NONE:
                ret.l = '';
                ret.r = '';
                break;
            case CarouselArrowTypes.SMALL:
                ret.l = '2039';
                ret.r = '203A';
                break;
            default:
            case CarouselArrowTypes.NORMAL:
                ret.l = '276E';
                ret.r = '276F';
                break;
            case CarouselArrowTypes.BOLD:
                ret.l = '2770';
                ret.r = '2771';
                break;
            case CarouselArrowTypes.NARROW:
                ret.l = '2329';
                ret.r = '232A';
                break;
        }
        return ret;
    };
    /**
     * Check if numeric value
     * @param value
     * @returns {boolean}
     */
    CarouselDirective.isNumeric = function (value) {
        return !isNaN(value - parseFloat(value));
    };
    CarouselDirective.animationSpeedDefault = 400; // in ms
    CarouselDirective.autoPlaySpeedDefault = 0; // in ms
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], CarouselDirective.prototype, "carousel", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number)
    ], CarouselDirective.prototype, "carouselSpeed", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], CarouselDirective.prototype, "carouselArrows", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], CarouselDirective.prototype, "carouselLabelOverlay", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number)
    ], CarouselDirective.prototype, "carouselAnimationSpeed", void 0);
    CarouselDirective = CarouselDirective_1 = __decorate([
        core_1.Directive({ selector: '[carousel]' }),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], CarouselDirective);
    return CarouselDirective;
    var CarouselDirective_1;
}());
exports.CarouselDirective = CarouselDirective;
var CarouselArrowTypes;
(function (CarouselArrowTypes) {
    CarouselArrowTypes[CarouselArrowTypes["NONE"] = 0] = "NONE";
    CarouselArrowTypes[CarouselArrowTypes["SMALL"] = 1] = "SMALL";
    CarouselArrowTypes[CarouselArrowTypes["NARROW"] = 2] = "NARROW";
    CarouselArrowTypes[CarouselArrowTypes["NORMAL"] = 3] = "NORMAL";
    CarouselArrowTypes[CarouselArrowTypes["BOLD"] = 4] = "BOLD";
})(CarouselArrowTypes || (CarouselArrowTypes = {}));
var CarouselDirections;
(function (CarouselDirections) {
    CarouselDirections[CarouselDirections["DIRECTION_LEFT"] = 0] = "DIRECTION_LEFT";
    CarouselDirections[CarouselDirections["DIRECTION_RIGHT"] = 1] = "DIRECTION_RIGHT";
})(CarouselDirections || (CarouselDirections = {}));
var CarouselSlide = /** @class */ (function () {
    function CarouselSlide() {
    }
    return CarouselSlide;
}());
exports.CarouselSlide = CarouselSlide;
